src/rules/no-invalid-label-refs.js

```diff
@@ -0,0 +1,140 @@
+/**
+ * @fileoverview Rule to prevent non-complaint link references.
+ * @author Nicholas C. Zakas
+ */
+
+//-----------------------------------------------------------------------------
+// Imports
+//-----------------------------------------------------------------------------
+
+import { findOffsets, illegalShorthandTailPattern } from "../util.js";
+
+//-----------------------------------------------------------------------------
+// Type Definitions
+//-----------------------------------------------------------------------------
+
+/** @typedef {import("unist").Position} Position */
+/** @typedef {import("mdast").Text} TextNode */
+/** @typedef {import("eslint").Rule.RuleModule} RuleModule */
+
+//-----------------------------------------------------------------------------
+// Helpers
+//-----------------------------------------------------------------------------
+
+// matches i.e., [foo][bar]
+const labelPattern = /\]\[([^\]]+)\]/u;
+
+
+/**
+ * Finds missing references in a node.
+ * @param {TextNode} node The node to check.
+ * @param {string} text The text of the node.
+ * @returns {Array<{label:string,position:Position}>} The missing references.
+ */
+function findIllegalLabelReferences(node, text) {
+
+    const invalid = [];
+    let startIndex = 0;
+
+    while (startIndex < node.value.length) {
+
+        const value = node.value.slice(startIndex);
+        const match = value.match(labelPattern);
+
+        if (!match) {
+            break;
+        }
+
+        if (!illegalShorthandTailPattern.test(match[0])) {
+            continue;
+        }
+
+        let columnStart = startIndex + match.index + 1;
+
+        // adding 1 to the index just in case we're in a ![] and need to skip the !.
+        const startFrom = node.position.start.offset + startIndex + 1;
+        const lastOpenBracket = text.lastIndexOf("[", startFrom);
+
+        if (lastOpenBracket === -1) {
+            startIndex += match.index + match[0].length;
+            continue;
+        }
+
+        const label = text.slice(lastOpenBracket, match.index + match[0].length).match(/!?\[([^\]]+)\]/u)?.[1];
+
+        columnStart -= label.length;
```

Comment:
Thanks, I'll take a look.

Reply:
This case is also still crashing the rule:

```js
        {
            code: "- - - [foo][ ]\n\n[foo]: http://foo.com",
            errors: [
                {
                    messageId: "invalidLabelRef",
                    data: { label: "foo" },
                    line: 1,
                    column: 8,
                    endLine: 1,
                    endColumn: 11
                }
            ]
        }
```

I think that something is off with the values of the arguments that get passed to `text.slice()`. The optional chaining `?.[1]` is probably not necessary, because `label` should not be undefined, or otherwise a `TypeError` will occur.